keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1053 @@
+-->
+
+No. It's a new API field, so no default behavior will be impacted.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes. If disabled, kube-apiserver will start rejecting Pod's mutation on `.spec.schedulingGates`.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Mutation on Pod's `.spec.schedulingGates` will be respected again.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+Yes, there are tests at unit and integration level.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+### Monitoring Requirements
```

Comment:
@Huang-Wei perhaps we can add a label to track those pods separately in our existing metrics; we have `active`, `backoff` and `unschedulable`; we could add `notReady`

Reply:
SG.

Reply:
I would prefer a label to specifically track gated for scheduling, which is not `notReady` (ready is a specific defined term, let's not overload it).  We should come up with a specific term used to refer to pods that have not closed all scheduling gates, that word should be clearly associated with the phase it is in, while ready is associated with a later phase.

Reply:
How about `waiting` / `paused` / `gated`? (with an implicit "scheduling" context as prefix)

Reply:
I would be ok with `notSchedulingReady`, but we should not use `Ready` by itself in any of our APIs / metrics because that could confuse users.

"Ready to schedule" reads ok, "scheduling readiness" is a bit less clear but still ok, but in constants / fields we can't say `pod ready` / `pod not ready` without the word `schedule` or `scheduling`.  I can't think of a better word right now (`appropriate for scheduling` sounds wierd, `schedulingGated` is awkward and doesn't negate well, `allowed` is too close to rbac or policy), but there may be a better word out there if we think about it.  For right now `scheduling readiness` and fully qualified variants are ok for alpha, and we should consider better wording for this phase.